src/middlewared/middlewared/plugins/virt/global.py

```diff
@@ -0,0 +1,372 @@
+
+        bridge = new['bridge'] or ''
+        if bridge not in await self.bridge_choices():
+            verrors.add(f'{schema_name}.bridge', 'Invalid bridge')
+
+        pool = new['pool'] or ''
+        if pool not in await self.pool_choices():
+            verrors.add(f'{schema_name}.pool', 'Invalid pool')
+
+    @api_method(VirtGlobalUpdateArgs, VirtGlobalUpdateResult)
+    @job()
+    async def do_update(self, job, data):
+        """
+        Update global virtualization settings.
+
+        `pool` which pool to use to store instances.
+        """
+        old = await self.config()
+
+        new = old.copy()
+        new.update(data)
+
+        verrors = ValidationErrors()
+        await self.validate(new, 'virt_global_update', verrors)
+        verrors.check()
+
+        # Not part of the database
+        new.pop('state')
+        new.pop('dataset')
+
+        await self.middleware.call(
+            'datastore.update', self._config.datastore,
+            new['id'], new,
+        )
+
+        job = await self.middleware.call('virt.global.setup')
+        await job.wait(raise_error=True)
+
+        return await self.config()
+
+    @api_method(VirtGlobalBridgeChoicesArgs, VirtGlobalBridgeChoicesResult, roles=['VIRT_GLOBAL_READ'])
+    async def bridge_choices(self):
+        """
+        Bridge choices for virtualization purposes.
+
+        Empty means it will be managed/created automatically.
+        """
+        choices = {'': 'Automatic'}
+        choices.update({
+            i['name']: i['name'] for i in await self.middleware.call('interface.query')
+            if i['type'] == 'BRIDGE'
+        })
+        return choices
+
+    @api_method(VirtGlobalPoolChoicesArgs, VirtGlobalPoolChoicesResult, roles=['VIRT_GLOBAL_READ'])
+    async def pool_choices(self):
+        """
+        Pool choices for virtualization purposes.
+        """
+        pools = {'': '[Disabled]'}
+        for p in (await self.middleware.call('zfs.pool.query_imported_fast')).values():
+            if p['name'] in BOOT_POOL_NAME_VALID:
+                continue
+            ds = await self.middleware.call(
+                'pool.dataset.get_instance_quick', p['name'], {'encryption': True},
+            )
+            if not ds['locked']:
+                pools[p['name']] = p['name']
+        return pools
+
+    @private
+    async def internal_interfaces(self):
+        return [INCUS_BRIDGE]
+
+    @private
+    async def get_default_profile(self):
+        result = await incus_call('1.0/profiles/default', 'get')
+        if result.get('status_code') != 200:
+            raise CallError(result.get('error'))
+        return result['metadata']
+
+    @api_method(VirtGlobalGetNetworkArgs, VirtGlobalGetNetworkResult, roles=['VIRT_GLOBAL_READ'])
+    async def get_network(self, name):
+        """
+        Details for the given network.
+        """
+        result = await incus_call(f'1.0/networks/{name}', 'get')
```

Comment:
This will fail if incus is not setup, we should have a safety net here of sorts so we fail appropriately i.e incus pool is not setup